data/transactions/logic/eval.go

```diff
@@ -2102,6 +2103,56 @@ func opGtxnsa(cx *evalContext) {
 	cx.stack[last] = sv
 }
 
+func opGaidImpl(cx *evalContext, groupIdx int, opName string) (sv stackValue, err error) {
+	if groupIdx >= len(cx.TxnGroup) {
+		err = fmt.Errorf("%s lookup TxnGroup[%d] but it only has %d", opName, groupIdx, len(cx.TxnGroup))
+		return
+	} else if groupIdx > cx.GroupIndex {
+		err = fmt.Errorf("%s can't get future creatable ID of txn with index %d", opName, groupIdx)
+		return
+	} else if groupIdx == cx.GroupIndex {
+		err = fmt.Errorf("can't use %s on self, use `global CurrentApplicationID` instead", opName)
```

Comment:
interesting... but why not ? i.e. I think that we should be able to do that, right ?
If we do decide to keep this - should we also add this check to the static analysis ?

Reply:
I don't think we should allow access to "self" with these kind of opcodes (`gload` and `gaid`).

We do allow it with `gtxn` even though it is "dumb", that is, it's an extra byte of code for no advantage over `txn`.  I think the reason we allow it is because it would be awfully strange to have the opcode allow access to earlier txns and later txns, but exclude self.

With these opcodes, that already can't look into higher slots, that reasoning breaks down.  We are definitely not going to let them look *past* self, so should we exclude self too?  It won't leave an unsightly hole this time, but it remains an ugly way to do something that is already allowed.

The argument for: maybe you want to loop from 0 to self, doing the same work on each transaction, including self.  My argument against: That's almost certainly an off-by-one bug that you didn't intend.  When txns use an opcode like this, they're trying to confirm that something happened in another transaction.  I can't imagine needing to do the same kind of examination on a bunch of transactions *including self*.

But I'm not sure we should add a static check.  We *can't* for `gaids` which reads the index from scratch.  And we can't at assembly time for either of them, since we don't know our index at assembly time.  The only possibility is to check the `gaid <slot>` case, at network submission time.  That will certainly fail, yes, but so would `gtxn 62`. Nonetheless, we allow it (at Check() time, and fail at Eval() time). I think we might want a comprehensive static checker, but I'm not sure it should be in algod.  It should be a tool to find bugs ahead of time, used by developers.

Reply:
> I would be in favor of a single >= check though, and write the error message so it simply says you can only look at previous slots in the group.

I'd reword the error message to state that `gaid` is only for looking at previous slots in the group, but I'd rather not leave the user hanging by not offering an alternative when there is an straightforward one available.